reactioncommerce/reaction
# Convert search modal wrapper to React

Resolves #3424  
Impact: **breaking**  
Type: **refactor**

## Important Note

~~This was branched off of `refactor-3495-mikemurray-react16` which has the update to React `16.2.0` See PR https://github.com/reactioncommerce/reaction/pull/3801 for more details.~~

~~This PR should be merged after or in place of https://github.com/reactioncommerce/reaction/pull/3801~~

https://github.com/reactioncommerce/reaction/pull/3801 Has been merged to master in release `1.8.2`.

## Issue

The search modal, while being all react, was being injected into the dom tree using Blaze and jQuery. This PR removes the dependency on Blaze and the search modal Blaze wrappers in favor of a React only solution.

## Solution

Used Portals in React 16 to inject the search modal in the DOM container `search-modal-root` outside of `react-root` to ensure the modal is not restricted by other containers. 

Portals allow for React components to live anywhere in the DOM tree while still being part of the _React tree_. This is especially useful for modals where a parent component may have an `overflow:hidden` that would make the modal not visible if rendered in that container.

More info on portals: https://reactjs.org/docs/portals.html

jQuery was removed as a dependency. It was used to auto select the search input on the modal show, as well as highlighting the active tab between "Products" and "Accounts" for admins. This has been refactored to the React way.

## Breaking changes

**Marked as breaking mainly because files were removed.**

Removed search modal Blaze wrapper component and imports.

The implementation of `core/ui-navbar` and `core/ui-search` is now using Portals in React 16 to add the search modal to the DOM tree.

## Testing

The search modal should be able to be opened, used, and closed by both admin and customers. Admins should additionally be able to search for accounts in the search modal.

Add as many products and users as needed to ensure the search modal works as expected.

**Test 1**
1. As a customer
2. Open search modal, and type in "basic" to see the basic reaction product
3. Close the modal

**Test 2**
1. As an admin
1. Open search modal, and type in "basic" to see the basic reaction product
1. Switch to the accounts tab and search for an account, or the admin account
1. Close the modal
